better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1044 @@
+functions, types, and modules, and why we try to keep them small.
+
+## Components and Abstraction
+
+Obviously, “knowing something about `sort`” is only possible and
+relevant because `sort` is a separate component with its own name. If
+we had injected the sorting implementation directly into our code,
+we'd have to reason about all of its details, and we'd need a comment
+to tell us that the implementation was sorting the elements.
+
+So it's a common fallacy that factoring out a component only makes
+sense if it's used more than once. The real criterion is whether you
+can identify a unit of *abstraction*—the third ingredient of Better
+Code.  If you can think of a simple and descriptive name for
+something, that's usually a good indicator that there's an abstraction
+waiting to be discovered.
+
+In fact, creating a component directly creates the opportunity for an
+additional use site: in testing.  The more code you test separately,
+the better it is for correctness.  The more you can use understandable
+abstractions, the simpler and clearer the use-sites become, which is
+better for correctness. Abstraction and correctness support one
+another in a virtuous cycle.  And abstraction is necessary for
+programming at scale: because we can't keep the whole program in our
+heads, we do what humans always do when faced with complexity: we
+break complicated problems into parts that can be understood in
+isolation.
+
+## Hoare Logic
+
+Programming by contract grew out of Hoare Logic, developed in 1969 by
+the British computer scientist and logician Tony Hoare.  While we
+won't be using Hoare Logic directly, we'll discuss it briefly to
+provide historical context, and because it is the source of crucial
+terminology.
+
+### Preconditions and Postconditions
+
+Hoare used this notation, called a “Hoare triple,”
+
+> {P}S{Q}
+
+which is an assertion that if **precondition** *P* is met, operation
+*S* establishes **postcondition** *Q*.
+
+For example:
+
+- if we start with `x == 2` (precondition), after `x += 1`, `x == 3` (postcondition):
+
+  > {x == 2}x+=1{x == 3}
+
+- if `x` is less than the maximum integer (precondition), after `x
+  += 1`, `x` is greater than the minimum integer (postcondition):
+
+  > {x < Int.max}x+=1{x > Int.min}
+
+What makes preconditions and postconditions useful for formal proofs
+is this composition rule:
+
+> {P}S{Q} ∧ {Q}T{R} ⇒ {P}S;T{R}
+
+Given two valid Hoare triples, if the postconditions of the first are
+preconditions of the second, we can form a new valid triple describing
+the effects of performing the operations in sequence.  The notation
+may look exotic, but its meaning should be familiar: this rule simply
+captures the reasoning we use to think informally about statements
+when we write one statement after the next.
+
+### Invariants
+
+Not all code runs in a straight line, though, so Hoare also gave us a
+tool for reasoning about loops.
+
+A **loop invariant** is a condition that holds before and after each
```

Comment:
Usually, this is phrased as "before entering the loop and after each iteration (including the last)".

Reply:
@sean-parent bump?